File: image-data-hash-py/tests/test_image_data_hash.py
import csv
import os
from io import BytesIO
from pathlib import Path

import pytest

from image_data_hash import (
    image_data_hash,
    image_data_hash_bytes,
    image_data_hash_stream,
)

DIRECTORY = Path(
    os.environ.get(
        "JPEG_HASH_FIXTURES", str(Path(__file__).parents[2] / "fixtures/generated")
    )
)
with (DIRECTORY / "manifest.tsv").open() as manifest:
    VECTORS = list(csv.DictReader(manifest, delimiter="\t"))
assert len(VECTORS) == 41, "Generate all compatibility fixtures before testing"


@pytest.mark.parametrize("row", VECTORS, ids=lambda row: row["file"])
@pytest.mark.parametrize("algorithm", ["md5", "sha256", "sha512"])
def test_matches_exiftool(row, algorithm):
    path = DIRECTORY / row["file"]
    assert image_data_hash(path, algorithm) == row[algorithm]
    assert image_data_hash_bytes(path.read_bytes(), algorithm) == row[algorithm]


def test_default_and_algorithm_aliases():
    path = DIRECTORY / VECTORS[0]["file"]
    assert image_data_hash(path) == VECTORS[0]["md5"]
    assert image_data_hash(path, "SHA-256") == VECTORS[0]["sha256"]
    assert image_data_hash(path, "SHA-512") == VECTORS[0]["sha512"]
    with pytest.raises(ValueError):
        image_data_hash(path, "sha1")


class ShortReader(BytesIO):
    def read(self, size=-1):
        return super().read(min(size, 1))


@pytest.mark.parametrize(
    "row", [row for row in VECTORS if row["file"].startswith("structural")]
)
def test_split_markers(row):
    source = ShortReader((DIRECTORY / row["file"]).read_bytes())
    assert image_data_hash_stream(source) == row["md5"]
    assert not source.closed


@pytest.mark.parametrize(
    "data",
    [
        b"",
        b"not jpeg",
        b"\xff\xd8",
        b"\xff\xd8\xff\xe1\x00",
        b"\xff\xd8\xff\xe1\x00\x01",
        b"\xff\xd8\xff\xd9",
        b"\xff\xd8\xff\xda\x00\x02abc\xff",
    ],
)
def test_invalid_input(data):
    with pytest.raises(ValueError):
        image_data_hash_bytes(data)

File: image-data-hash-r/test_image_data_hash.R
library(testthat)
source("R/image_data_hash.R")

directory <- Sys.getenv("JPEG_HASH_FIXTURES", "../fixtures/generated")
manifest <- read.delim(file.path(directory, "manifest.tsv"), stringsAsFactors = FALSE)

test_that("all fixtures match ExifTool for every algorithm", {
  expect_equal(nrow(manifest), 41L)
  for (i in seq_len(nrow(manifest))) {
    path <- file.path(directory, manifest$file[i])
    data <- readBin(path, "raw", n = file.info(path)$size)
    expect_equal(image_data_hash_raw(data), manifest$md5[i])
    for (algorithm in c("md5", "sha256", "sha512")) {
      expect_equal(image_data_hash(path, algorithm), manifest[[algorithm]][i], info = manifest$file[i])
      expect_equal(image_data_hash_raw(data, algorithm), manifest[[algorithm]][i], info = manifest$file[i])
    }
  }
})

test_that("algorithm aliases and invalid inputs", {
  path <- file.path(directory, manifest$file[1])
  expect_equal(image_data_hash(path, "SHA-256"), manifest$sha256[1])
  expect_equal(image_data_hash(path, "SHA-512"), manifest$sha512[1])
  expect_error(image_data_hash(path, "sha1"))
  expect_error(image_data_hash_raw("not bytes"))
  for (bytes in list(
    integer(), c(1, 2, 3, 4), c(255, 216), c(255, 216, 255, 217),
    c(255, 216, 255, 225, 0, 1), c(255, 216, 255, 218, 0, 2, 255)
  )) {
    expect_error(image_data_hash_raw(as.raw(bytes)))
  }
})
